### PR TITLE
Enabling OAuth Authentication  for secondary user store retrieval via a back end service wrapper   

### PR DESCRIPTION
**Pull request by Kamidu**

### CODE REVIEW

File: modules/apps/jaggery/mdm/src/mdm/config/config.json [hunk 1]
```diff
@@ -5,6 +5,10 @@
     "httpURL" : "%http.ip%",
     "iOSConfigRoot" : "%https.ip%/ios-enrollment/",
     "iOSAPIRoot" : "%https.ip%/ios/",
+	"carbonServer":"%https.ip%",
```
Comment (dulichan) [inline modules/apps/jaggery/mdm/src/mdm/config/config.json]: Change this to dynamic client registration service

File: modules/apps/jaggery/mdm/src/mdm/modules/api-wrapper-util.js [hunk 2]
```diff
@@ -22,27 +22,32 @@ var apiWrapperUtil = function () {
...
-            tokenPair = tokenUtil.getTokenWithPasswordGrantType(properties.username, properties.password, clientData.clientId, clientData.clientSecret);
-        } else if (type == "saml") {log.error(clientData);
-            var scope = "admin";log.error("samltoken >>>>>>>>>> " + properties.samlToken);
-            tokenPair = tokenUtil.getTokenWithSAMLGrantType(properties.samlToken, clientData.clientId, clientData.clientSecret, scope);
+            tokenPair =
+                tokenUtil.getTokenWithPasswordGrantType(properties.username, properties.password, encodedClientKeys);
+        } else if (type == "saml") {
+            log.error(clientData);
+            var scope = "admin";
+            log.error("samltoken >>>>>>>>>> " + properties.samlToken);
```
Comment (dulichan) [inline modules/apps/jaggery/mdm/src/mdm/modules/api-wrapper-util.js]: Remove this log and print the user name

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 3]
```diff
@@ -420,25 +437,12 @@ var userModule = function () {
         }
         try {
             utility.startTenantFlow(carbonUser);
-
+            log.error("here")
```
Comment (dulichan) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Make this log meaningful

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 4]
```diff
@@ -420,25 +437,12 @@ var userModule = function () {
...
-                // generate response
-                response["status"] = "success";
-                response["content"] = userList;
-                return response;
-            } else {
-                // generate response
-                response["status"] = "error";
-                return response;
-            }
+            log.error("hmmmm");
```
Comment (dulichan) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: -_-

File: modules/apps/jaggery/mdm/src/mdm/modules/api-wrapper-util.js (hunk 2 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/api-wrapper-util.js]: Okay as per the discussion i had with you i have removed all the log and log errors in api-wrapper-util.js which was added previously .

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 3 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: This was a fault due to that all the committed changes has been reverted and all the local changes has been re based.Issue has been fixed as soon as encountered.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 4 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: This was a fault due to that all the committed changes has been reverted and all the local changes has been re based.Issue has been fixed as soon as encountered.

File: modules/apps/jaggery/mdm/src/mdm/config/config.json (hunk 1 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/config/config.json]: I have added a separate entry for dynamic client registration and change the current carbon server entry as adminService .

**Comment by dulichan** [comment]
Locally tested with a patch, no build conflicts or detrimental changes.

File: modules/apps/jaggery/mdm/src/mdm/api/user-api.jag [hunk 5]
```diff
@@ -40,7 +40,7 @@ if (uriMatcher.match("/{context}/api/user/login/")) {
                 log.debug("User Logged In : " + user);
             }
             utility.insertAppPermissions(userModule, "login");
-            //apiWrapperUtil.setupAccessTokenPair("password", {"username": username, "password": password});
+            apiWrapperUtil.setupAccessTokenPair("password", {"username": username, "password": password});
             var permissions = userModule.getUIPermissions();
             if (permissions.VIEW_DASHBOARD){
```
Comment (harshanL) [inline modules/apps/jaggery/mdm/src/mdm/api/user-api.jag]: Fix formatting issues.

Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/api/user-api.jag]: These formatting issues has been fixed in a recent PR

File: modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        methodObject.addRequestHeader(header);
+        if (IS_OAUTH_ENABLED) {
+            var accessToken = session.get(constants.ACCESS_TOKEN_PAIR_IDENTIFIER).accessToken;
+            if (!(!accessToken.trim())) {
+                header = new Header();
+                header.setName(constants.AUTHORIZATION_HEADER);
+                header.setValue(constants.BEARER_PREFIX + accessToken);
+                methodObject.addRequestHeader(header);
+            }
+
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js]: Remove unnecessary new line.

File: modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        var ws = require('ws');
+        var wsRequest = new ws.WSRequest();
+        var options = [];
+        if (IS_OAUTH_ENABLED) {
+            var accessToken = session.get(constants.ACCESS_TOKEN_PAIR_IDENTIFIER).accessToken;
+            if (!(!accessToken.trim())) {
+                var headers = [{name: constants.AUTHORIZATION_HEADER, value: constants.BEARER_PREFIX + accessToken}];
+                options.HTTPHeaders = headers;
+            }
+
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js]: Remove unnecessary newline.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 8]
```diff
@@ -62,26 +63,38 @@ var userModule = function () {
...
+                                                           function (responsePayload) {
+                                                               var response = {};
+                                                               response.content = responsePayload;
+                                                               response.status = "error";
+                                                               return response;
+                                                           });
+                return response;
+                break;
+            case constants.HTTP_POST:
+                //todo
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Why TODOs?

File: modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js (hunk 7 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js]: These formatting issues has been fixed in a recent PR

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 8 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: This Method only supports GET request currently.this was added to extend the method when needed.As per the discussion had offline recommended changes are applied.
Recommended fixes can be found in PR[1]

[1]. https://github.com/wso2/product-mdm/pull/166

File: modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        if (IS_OAUTH_ENABLED) {
+            var accessToken = session.get(constants.ACCESS_TOKEN_PAIR_IDENTIFIER).accessToken;
+            if (!(!accessToken.trim())) {
+                xmlHttpRequest.setRequestHeader(
+                    constants.AUTHORIZATION_HEADER, constants.BEARER_PREFIX + accessToken);
+            }
+        }
+        xmlHttpRequest.send(stringify(payload));
+        if (xmlHttpRequest.status == 200) {
+            log.error(parse(xmlHttpRequest["responseText"]));
```
Comment (dakshika) [inline modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js]: status 200 is success call, why we need to push log.error to console?

Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/backendServiceInvoker.js]: This issue were corrected in a recent PR.

**Pull request merged**